typing/path.ml

```diff
@@ -88,6 +94,52 @@ let rec scope = function
   | Pdot(p, _) | Pextra_ty (p, _) -> scope p
   | Papply(p1, p2) -> Int.max (scope p1) (scope p2)
 
+let rec contains id = function
+    Pident id' -> Ident.same id' id
+  | Pdot(p, _) | Pextra_ty(p, _) -> contains id p
+  | Papply(p1, p2) -> contains id p1 || contains id p2
+
+let subst id_map p =
+  let exception Unchanged in
+  let rec aux = function
+  | Pident id ->
+    begin match List.find (fun (i, _) -> Ident.same i id) id_map with
```

Comment:
I am slightly nervous about the O(n) cost of substitutions expressed in this way. There could be a comment somewhere that explains why we believe that these lists are going to remain small. (If I understand `Ctype.nondep_type_rec` and `copy_sep`, maybe this is bounded by the maximal nesting of module-dependent arrow types in the syntax?) Or we could use `Map.Make(Ident)` and the potential problem would go away?

Reply:
I remain uncomfortable with the linear traversal implied by lists being passed around everywhere. On the other hand, the current code is simple and presumably correct. I propose to leave the change of data structure for a follow-up PR, and @samsa1 seems okay with that approach.